In [1]:
!pip install --upgrade sentence-transformers transformers 

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 470.2/470.2 kB 17.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 107.9 MB/s eta 0:00:0000:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.6 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 80.2 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 71.4 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 41.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.5 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 7.9 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 29.7 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 9.5 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

### ⚙️ **Compose Data Frames With All Post Descriptions and User Info** ⚙️

In [2]:
import json
import pandas as pd
import numpy as np

In [3]:
with open("/kaggle/input/o4u-messages/o4u_messages_Jun_07_2025.json", "r", encoding="utf-8") as f:
    posts = json.load(f)

with open("/kaggle/input/o4u-messages/o4u_preprocessed_messages_Jun_07_2025.json", "r", encoding="utf-8") as f:
    preprocessed_posts = json.load(f)

In [4]:
df_posts = pd.DataFrame(
    np.column_stack((list(range(len(posts))), posts, preprocessed_posts, [None] * len(posts))),  
    columns=['post_id', 'post', 'preprocessed_post', 'embedding']
)

df_posts.head()

post_id                                               post  \
0       0  Dear students,\n\nThis channel advertises mino...   
1       1  📣Hi there!\n\nStudent Affairs is urgently look...   
2       2  📣Hi there! Want any of these?\n\nStudent Affai...   
3       3  Bonjour! Ça va?\nС'est la vie.. \nCroissant.\n...   
4       4  📣On December 14, comedian Vladimir Marconi arr...   

                                   preprocessed_post embedding  
0  Dear students, This channel advertises minor e...      None  
1  Hi there! Student Affairs is urgently looking ...      None  
2  Hi there! Want any of these? Student Affairs a...      None  
3  Bonjour! Ça va? С'est la vie.. Croissant. If t...      None  
4  On December 14, comedian Vladimir Marconi arri...      None

In [5]:
df_users = pd.DataFrame(
    columns=['user_id', 'description', 'post_id', 'commented', 'upvoted', 'attended', 'reactions']
)

data_types = {
    'user_id': np.int32,
    'description': 'string',
    'post_id': np.int32,
    'commented': pd.BooleanDtype(),
    'upvoted': pd.BooleanDtype(),
    'attended': pd.BooleanDtype(),
    'reactions': 'string'
}

df_users = df_users.astype(data_types)

df_users.head()

Empty DataFrame
Columns: [user_id, description, post_id, commented, upvoted, attended, reactions]
Index: []

### 👤 **Create Users With Different Preferences** 👤

In [6]:
import random

In [7]:
user_personas = [
    {
        "user_id": 1,
        "description": "Competitive Coder",
        "positive_posts": [10, 19, 67, 71, 72, 75, 118, 181, 242, 316],
        "negative_posts": [5, 14, 18, 22, 34, 45, 51, 150, 209, 216]
    },
    {
        "user_id": 2,
        "description": "Creative Artist",
        "positive_posts": [14, 18, 34, 45, 51, 57, 127, 209, 251, 263],
        "negative_posts": [10, 11, 13, 19, 20, 72, 112, 119, 148, 154]
    },
    {
        "user_id": 3,
        "description": "Career-Focused Student",
        "positive_posts": [11, 20, 29, 50, 66, 69, 112, 147, 221, 273],
        "negative_posts": [5, 18, 35, 45, 150, 184, 208, 216, 240, 258]
    },
    {
        "user_id": 4,
        "description": "Community Volunteer",
        "positive_posts": [1, 2, 7, 16, 22, 24, 26, 108, 217, 243],
        "negative_posts": [10, 13, 19, 67, 72, 75, 119, 200, 215, 242]
    },
    {
        "user_id": 5,
        "description": "Robotics & Hardware Tinkerer",
        "positive_posts": [13, 17, 28, 43, 62, 111, 115, 127, 222, 309],
        "negative_posts": [3, 5, 14, 22, 34, 45, 51, 76, 114, 251]
    }
]

positive_reactions = ["👍", "❤️", "🔥"]
negative_reactions = ["👎", "😐"]

new_user_data = []
for user in user_personas:
    # Generate positive interactions
    for post_id in user["positive_posts"]:
        reaction_data = {
            "user_id": user["user_id"],
            "description": user["description"],
            "post_id": post_id,
            "commented": random.choice([True, False, False]),  # 33% chance
            "upvoted": True,
            "attended": random.choice([True, False]),  # 50% chance
            "reactions": random.choice(positive_reactions)
        }
        new_user_data.append(reaction_data)
        
    # Generate negative interactions
    for post_id in user["negative_posts"]:
        reaction_data = {
            "user_id": user["user_id"],
            "description": user["description"],
            "post_id": post_id,
            "commented": random.choice([True, False, False, False]),  # 25% chance
            "upvoted": False,
            "attended": False,
            "reactions": random.choice(negative_reactions)
        }
        new_user_data.append(reaction_data)

if new_user_data:
    df_new_users = pd.DataFrame(new_user_data)
    df_users = pd.concat([df_users, df_new_users], ignore_index=True).astype(data_types)

display(df_users)

user_id                   description  post_id  commented  upvoted  \
0         1             Competitive Coder       10      False     True   
1         1             Competitive Coder       19       True     True   
2         1             Competitive Coder       67       True     True   
3         1             Competitive Coder       71       True     True   
4         1             Competitive Coder       72      False     True   
..      ...                           ...      ...        ...      ...   
95        5  Robotics & Hardware Tinkerer       45      False    False   
96        5  Robotics & Hardware Tinkerer       51      False    False   
97        5  Robotics & Hardware Tinkerer       76      False    False   
98        5  Robotics & Hardware Tinkerer      114      False    False   
99        5  Robotics & Hardware Tinkerer      251       True    False   

    attended reactions  
0       True        ❤️  
1      False         🔥  
2      False         🔥  
3      False        ❤️  
4       True         🔥  
..       ...       ...  
95     False         👎  
96     False         😐  
97     False         😐  
98     False         👎  
99     False         👎  

[100 rows x 7 columns]

In [8]:
%%capture cell
# --- User 6 ---
user_6_desc = 'Creative Person'
'''
This user loves all events that are related to art, music, and photography. 

They attend almost every one.

They comment on events and react to them reasonably frequently, as well as upvoting/downvoting them.
'''
# Creative events
df_users.loc[len(df_users)] = [6, user_6_desc, 2499, True, True, True, '😇']
df_users.loc[len(df_users)] = [6, user_6_desc, 2403, False, None, True, '🥰']
df_users.loc[len(df_users)] = [6, user_6_desc, 2064, True, True, True, '❤️']
df_users.loc[len(df_users)] = [6, user_6_desc, 2187, False, None, True, '🥰']
df_users.loc[len(df_users)] = [6, user_6_desc, 1999, False, True, True, '❤️']
df_users.loc[len(df_users)] = [6, user_6_desc, 2634, True, True, True, '😇']
df_users.loc[len(df_users)] = [6, user_6_desc, 1705, True, True, True, '👍']
df_users.loc[len(df_users)] = [6, user_6_desc, 1244, False, None, True, '🔥']
df_users.loc[len(df_users)] = [6, user_6_desc, 872, True, None, False, '😇']
df_users.loc[len(df_users)] = [6, user_6_desc, 834, False, True, True, '🥰']
df_users.loc[len(df_users)] = [6, user_6_desc, 364, True, None, True, '❤️']
df_users.loc[len(df_users)] = [6, user_6_desc, 2699, False, True, True, '👍']
df_users.loc[len(df_users)] = [6, user_6_desc, 2206, True, True, False, '🥰']


# --- User 7 ---
user_7_desc = 'Sports User'
'''
This user loves all sports events. 

It attends almost every one and takes a role of a participant or a volunteer.

However, it rarely reacts to or upvotes posts, nor does it ever comment on them.
'''
# Sport events
df_users.loc[len(df_users)] = [7, user_7_desc, 2418, True, None, True, '👍']
df_users.loc[len(df_users)] = [7, user_7_desc, 2045, False, True, True, '👍']
df_users.loc[len(df_users)] = [7, user_7_desc, 639, False, None, True, '👍']
df_users.loc[len(df_users)] = [7, user_7_desc, 385, False, None, True, '❤️']
df_users.loc[len(df_users)] = [7, user_7_desc, 1265, False, True, True, '🔥']
df_users.loc[len(df_users)] = [7, user_7_desc, 155, False, True, False, '🔥']
df_users.loc[len(df_users)] = [7, user_7_desc, 52, True, None, False, '😱']
df_users.loc[len(df_users)] = [7, user_7_desc, 1982, False, None, True, '❤️']
df_users.loc[len(df_users)] = [7, user_7_desc, 885, False, None, True, '🔥']
df_users.loc[len(df_users)] = [7, user_7_desc, 2738, False, None, True, '❤️']
df_users.loc[len(df_users)] = [7, user_7_desc, 1487, False, None, True, '🔥']
# Some other events (quizzes, hackathons)
df_users.loc[len(df_users)] = [7, user_7_desc, 1258, True, None, True, '🥰']
df_users.loc[len(df_users)] = [7, user_7_desc, 2481, False, True, True, '🥰']
df_users.loc[len(df_users)] = [7, user_7_desc, 2035, False, None, True, '🔥']
df_users.loc[len(df_users)] = [7, user_7_desc, 2823, False, None, True, '🔥']
df_users.loc[len(df_users)] = [7, user_7_desc, 2775, False, None, True, '🥰']


# --- User 8 ---
user_8_desc = 'Mindful Person'
'''
This user loves attending events for girls and enjoys psychology sessions. 

However, they do not enjoy events such as quizzes and hackathons.

They often comment on events and frequently react and upvote/downvote them.
'''
# Events for girls
df_users.loc[len(df_users)] = [8, user_8_desc, 2620, True, True, True, '🥰']
df_users.loc[len(df_users)] = [8, user_8_desc, 2460, False, None, True, '😇']
df_users.loc[len(df_users)] = [8, user_8_desc, 2466, True, None, False, '🤯']
df_users.loc[len(df_users)] = [8, user_8_desc, 2243, True, True, True, '❤️']
df_users.loc[len(df_users)] = [8, user_8_desc, 1862, False, True, True, '😱']
df_users.loc[len(df_users)] = [8, user_8_desc, 1117, False, True, False, '❤️']
df_users.loc[len(df_users)] = [8, user_8_desc, 341, True, None, True, '😭']
df_users.loc[len(df_users)] = [8, user_8_desc, 1768, True, True, True, '😇']
df_users.loc[len(df_users)] = [8, user_8_desc, 1950, True, True, False, '🤯']
# Psychological events
df_users.loc[len(df_users)] = [8, user_8_desc, 2812, True, True, False, '🥰']
df_users.loc[len(df_users)] = [8, user_8_desc, 2804, True, None, True, '❤️']
df_users.loc[len(df_users)] = [8, user_8_desc, 2797, False, True, True, '👍']
df_users.loc[len(df_users)] = [8, user_8_desc, 2306, True, True, False, '❤️']
df_users.loc[len(df_users)] = [8, user_8_desc, 2393, True, True, True, '🥰']
df_users.loc[len(df_users)] = [8, user_8_desc, 2109, True, True, True, '🥰']
# Some other events (quizzes, hackathons)
df_users.loc[len(df_users)] = [8, user_8_desc, 2650, True, False, False, '😭']
df_users.loc[len(df_users)] = [8, user_8_desc, 2625, True, False, False, '👎']
df_users.loc[len(df_users)] = [8, user_8_desc, 2477, False, False, False, '😭']
df_users.loc[len(df_users)] = [8, user_8_desc, 2452, True, None, False, '👎']
df_users.loc[len(df_users)] = [8, user_8_desc, 1633, False, False, False, '😭']
df_users.loc[len(df_users)] = [8, user_8_desc, 2602, False, False, False, '👎']
df_users.loc[len(df_users)] = [8, user_8_desc, 2444, True, None, False, '😭']
df_users.loc[len(df_users)] = [8, user_8_desc, 2381, False, False, False, '😭']

# --- User 9 ---
user_9_desc = 'Activist'
'''
The focus of this user is on developing soft skills, career development and volunteering.

However, it does not like psychological events.

They comment on events and react to them reasonably frequently, as well as upvoting/downvoting them.
'''
# Career
df_users.loc[len(df_users)] = [9, user_9_desc, 2749, False, None, False, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 2677, True, False, True, '👍']
df_users.loc[len(df_users)] = [9, user_9_desc, 2395, False, True, True, '👍']
df_users.loc[len(df_users)] = [9, user_9_desc, 2325, True, True, False, '🥰']
df_users.loc[len(df_users)] = [9, user_9_desc, 2029, True, False, True, '😱']
# Volunteering
df_users.loc[len(df_users)] = [9, user_9_desc, 2725, False, True, False, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 2545, True, False, True, '🥰']
df_users.loc[len(df_users)] = [9, user_9_desc, 2513, True, False, False, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 1666, False, True, True, '🥰']
df_users.loc[len(df_users)] = [9, user_9_desc, 2343, True, None, True, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 2270, False, True, False, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 2160, True, False, True, '❤️']
df_users.loc[len(df_users)] = [9, user_9_desc, 2098, True, None, False, '😱']
df_users.loc[len(df_users)] = [9, user_9_desc, 2829, False, None, True, '🥰']
df_users.loc[len(df_users)] = [9, user_9_desc, 2717, True, True, False, '🥰']
df_users.loc[len(df_users)] = [9, user_9_desc, 2158, False, True, True, '😱']
# Psychological events
df_users.loc[len(df_users)] = [9, user_9_desc, 2393, True, False, False, '👎']
df_users.loc[len(df_users)] = [9, user_9_desc, 2109, False, False, False, '👎']
df_users.loc[len(df_users)] = [9, user_9_desc, 2804, True, False, False, '👎']
df_users.loc[len(df_users)] = [9, user_9_desc, 2797, False, False, False, '👎']


# --- User 10 ---
user_10_desc = 'Club & Party Person'
'''
This user likes stand-up events and parties.

However, they especially dislike festivals.

They comment on events and react to them reasonably frequently, as well as upvoting/downvoting them.
'''
# Standups & Parties
df_users.loc[len(df_users)] = [10, user_10_desc, 1915, True, True, True, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 1426, False, None, True, '👍']
df_users.loc[len(df_users)] = [10, user_10_desc, 1394, False, False, False, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 1022, True, True, True, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 877, False, None, False, '❤️']
df_users.loc[len(df_users)] = [10, user_10_desc, 2793, True, False, False, '❤️']
df_users.loc[len(df_users)] = [10, user_10_desc, 618, False, True, True, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 231, True, True, False, '👍']
df_users.loc[len(df_users)] = [10, user_10_desc, 1488, False, None, True, '❤️']
df_users.loc[len(df_users)] = [10, user_10_desc, 2035, False, True, False, '❤️']
df_users.loc[len(df_users)] = [10, user_10_desc, 2081, True, False, True, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 2657, False, True, False, '🔥']
df_users.loc[len(df_users)] = [10, user_10_desc, 2819, True, False, True, '🔥']
# Festivals
df_users.loc[len(df_users)] = [10, user_10_desc, 2799, False, None, False, '😭']
df_users.loc[len(df_users)] = [10, user_10_desc, 2658, True, True, False, '👎']
df_users.loc[len(df_users)] = [10, user_10_desc, 2043, False, False, False, '👎']
df_users.loc[len(df_users)] = [10, user_10_desc, 1843, True, True, False, '😭']
df_users.loc[len(df_users)] = [10, user_10_desc, 1574, True, True, False, '👎']
df_users.loc[len(df_users)] = [10, user_10_desc, 1205, False, False, False, '👎']
df_users.loc[len(df_users)] = [10, user_10_desc, 994, False, True, False, '👎']

new_posts = {
    1: [1542, 368, 818, 2294, 1519, 69, 976, 1918, 42, 188],
    2: [1766, 56, 1201, 2100, 1901, 578, 2521, 2520, 2678, 2516],
    3: [1413, 466, 40, 410, 193, 367, 2046, 2841, 1882, 597],
    4: [1732, 1569, 1443, 1477, 320, 650, 2646, 955, 573, 1481],
    5: [1713, 585, 2079, 1930, 1959, 2211, 986, 2103, 119, 1985],
    6: [1454, 438, 390, 824, 1234, 528, 2109, 1228, 536, 2027],
    7: [774, 470, 779, 2662, 1215, 183, 1425, 144, 2579, 812],
    8: [2461, 2471, 2828, 47, 224, 2469, 824, 256, 2027, 2554],
    9: [731, 587, 2399, 2398, 596, 561, 115, 94, 192, 1032],
    10: [1858, 1193, 1866, 1270, 876, 1501, 1351, 1789, 1128, 1929]
}

user_descriptions = {
    1: "Competitive Coder",
    2: "Creative Artist",
    3: "Career-Focused Student",
    4: "Community Volunteer",
    5: "Robotics & Hardware Tinkerer",
    6: "Creative Person",
    7: "Sports User",
    8: "Mindful Person",
    9: "Activist",
    10: "Club & Party Person"
}

positive_reactions = ["👍", "❤️", "🔥"]

new_user_data = []
for user_id, post_ids in new_posts.items():
    for post_id in post_ids:
        reaction_data = {
            "user_id": user_id,
            "description": user_descriptions[user_id],
            "post_id": post_id,
            "commented": random.choice([True, False, False]),  # 33% chance
            "upvoted": True,
            "attended": random.choice([True, False]),  # 50% chance
            "reactions": random.choice(positive_reactions)
        }
        new_user_data.append(reaction_data)

if new_user_data:
    df_new_users = pd.DataFrame(new_user_data)
    df_users = pd.concat([df_users, df_new_users], ignore_index=True).astype(data_types)

In [9]:
def get_user_posts_info(user_id: int) -> pd.DataFrame:
    """
    Lists the entries from `df_users` with `df_posts["preprocessed_post"]` appended to it.
    Sorts them by `upvoted` (descending) and then `post_id` (ascending).

    Args:
        - user_id (int): The user ID whose posts you want to view.
    
    Returns:
        - pd.DataFrame
    """

    return pd.merge(
        left=df_users[df_users["user_id"] == user_id],
        right=df_posts[["post_id", "preprocessed_post"]],
        on="post_id",
        how="left"
    ).sort_values(by=["upvoted", "post_id"], ascending=[False, True])

In [10]:
for user_id in range(1, df_users["user_id"].nunique() + 1):
    display(get_user_posts_info(user_id=user_id))

user_id        description post_id  commented  upvoted  attended  \
0         1  Competitive Coder      10      False     True      True   
1         1  Competitive Coder      19       True     True     False   
2         1  Competitive Coder      67       True     True     False   
3         1  Competitive Coder      71       True     True     False   
4         1  Competitive Coder      72      False     True      True   
5         1  Competitive Coder      75      False     True     False   
6         1  Competitive Coder     118      False     True      True   
7         1  Competitive Coder     181      False     True      True   
8         1  Competitive Coder     242       True     True     False   
9         1  Competitive Coder     316      False     True     False   
10        1  Competitive Coder       5       True    False     False   
11        1  Competitive Coder      14      False    False     False   
12        1  Competitive Coder      18      False    False     False   
13        1  Competitive Coder      22      False    False     False   
14        1  Competitive Coder      34       True    False     False   
15        1  Competitive Coder      45       True    False     False   
16        1  Competitive Coder      51      False    False     False   
17        1  Competitive Coder     150       True    False     False   
18        1  Competitive Coder     209      False    False     False   
19        1  Competitive Coder     216       True    False     False   

   reactions                                  preprocessed_post  
0         ❤️  Want to upgrade your programming skills? We ar...  
1          🔥  "Program the Future" contest from GS Labs rese...  
2          🔥  AMC Makeathon 2020 - Develop applications to s...  
3         ❤️  Blockchain Gamedev Meetup Videogames in the Bl...  
4          🔥  Internal ICPC tournament at Innopolis Universi...  
5          👍  REGISTRATION for the BREAKPOINT20 forum is OPE...  
6          🔥  World AI Data Challenge Online Hackathon 4-5 J...  
7          🔥  ICPC 2020 - Registration Reminder We remind yo...  
8         ❤️  Accounts Chamber of Russian Federation with Ge...  
9         ❤️  Registration open for Game Development in Unit...  
10         😐  Ho-Ho-Ho! SECRET SANTA IS COMING SOON! Secret ...  
11         😐  ART- Therapy course! Dear students! We continu...  
12         👎  Why to stay moody when you can shake your boot...  
13         👎  MEGA shopping centre is organizing a charitabl...  
14         😐  Japanese language courses at Anime Club! The f...  
15         👎  Poetry evening: date and location change! 16 F...  
16         😐  SPRING BALL 2020 When: 2nd April 19:00 What: E...  
17         😐  Announcing: Skateboard Club! We are pleased to...  
18         👎  reminder Dear friends! We are waiting for you ...  
19         👎  Just Dance! Break Dance club now has more to o...

user_id      description post_id  commented  upvoted  attended reactions  \
0         2  Creative Artist      14      False     True      True         👍   
1         2  Creative Artist      18      False     True      True         🔥   
2         2  Creative Artist      34       True     True      True         🔥   
3         2  Creative Artist      45       True     True      True         🔥   
4         2  Creative Artist      51      False     True      True         🔥   
5         2  Creative Artist      57       True     True      True         🔥   
6         2  Creative Artist     127      False     True     False        ❤️   
7         2  Creative Artist     209      False     True     False        ❤️   
8         2  Creative Artist     251       True     True      True         🔥   
9         2  Creative Artist     263      False     True     False         🔥   
10        2  Creative Artist      10       True    False     False         😐   
11        2  Creative Artist      11      False    False     False         😐   
12        2  Creative Artist      13       True    False     False         👎   
13        2  Creative Artist      19       True    False     False         😐   
14        2  Creative Artist      20      False    False     False         👎   
15        2  Creative Artist      72       True    False     False         😐   
16        2  Creative Artist     112      False    False     False         😐   
17        2  Creative Artist     119      False    False     False         😐   
18        2  Creative Artist     148      False    False     False         👎   
19        2  Creative Artist     154      False    False     False         😐   

                                    preprocessed_post  
0   ART- Therapy course! Dear students! We continu...  
1   Why to stay moody when you can shake your boot...  
2   Japanese language courses at Anime Club! The f...  
3   Poetry evening: date and location change! 16 F...  
4   SPRING BALL 2020 When: 2nd April 19:00 What: E...  
5   Student Club Night is starting at 19:30! See y...  
6   ARVR Hackathon in arts culture at Kazan State ...  
7   reminder Dear friends! We are waiting for you ...  
8   We invite you to the series of lectures on the...  
9                    Join Student Online Concert now:  
10  Want to upgrade your programming skills? We ar...  
11  We invite you to the Soft Skills School, which...  
12  APPLY for the Robotic projects until December ...  
13  "Program the Future" contest from GS Labs rese...  
14  Representatives of Молодежный Кадровый Потенци...  
15  Internal ICPC tournament at Innopolis Universi...  
16  TOP Vacancy (paid) IU are searching for a Web ...  
17  Rosneft IT-competition Marathon - Machine Lear...  
18  Speed Reading and Comprehension meetings Peopl...  
19      Application Form for INNOSTAR SCHOOL TEACHERS

user_id             description post_id  commented  upvoted  attended  \
0         3  Career-Focused Student      11       True     True     False   
1         3  Career-Focused Student      20      False     True      True   
2         3  Career-Focused Student      29       True     True      True   
3         3  Career-Focused Student      50      False     True     False   
4         3  Career-Focused Student      66      False     True      True   
5         3  Career-Focused Student      69      False     True     False   
6         3  Career-Focused Student     112       True     True     False   
7         3  Career-Focused Student     147       True     True      True   
8         3  Career-Focused Student     221      False     True     False   
9         3  Career-Focused Student     273      False     True      True   
10        3  Career-Focused Student       5       True    False     False   
11        3  Career-Focused Student      18      False    False     False   
12        3  Career-Focused Student      35      False    False     False   
13        3  Career-Focused Student      45      False    False     False   
14        3  Career-Focused Student     150      False    False     False   
15        3  Career-Focused Student     184      False    False     False   
16        3  Career-Focused Student     208      False    False     False   
17        3  Career-Focused Student     216      False    False     False   
18        3  Career-Focused Student     240      False    False     False   
19        3  Career-Focused Student     258      False    False     False   

   reactions                                  preprocessed_post  
0          👍  We invite you to the Soft Skills School, which...  
1          👍  Representatives of Молодежный Кадровый Потенци...  
2          👍  Student Affairs Office is looking for a part-t...  
3          👍  Dear Students, the International Relations Off...  
4          👍  X Republic Youth Forum Our Tatarstan Education...  
5         ❤️  World AIData Challenge International Project T...  
6         ❤️  TOP Vacancy (paid) IU are searching for a Web ...  
7         ❤️  Maxim Mikheenko Master Class You will find out...  
8          🔥  Acceleration and educational intensive program...  
9         ❤️  Opportunity for internship topics, course proj...  
10         👎  Ho-Ho-Ho! SECRET SANTA IS COMING SOON! Secret ...  
11         👎  Why to stay moody when you can shake your boot...  
12         👎  Board Games tournament with Hobby Games from K...  
13         😐  Poetry evening: date and location change! 16 F...  
14         😐  Announcing: Skateboard Club! We are pleased to...  
15         👎  IU Developers Community presentation - today 1...  
16         👎  Do you have an idea that might change student ...  
17         👎  Just Dance! Break Dance club now has more to o...  
18         😐  Armwrestling Tournament November 21st at 9am I...  
19         👎  Those students who are looking for a Junior De...

user_id          description post_id  commented  upvoted  attended  \
0         4  Community Volunteer       1      False     True      True   
1         4  Community Volunteer       2       True     True     False   
2         4  Community Volunteer       7      False     True     False   
3         4  Community Volunteer      16      False     True      True   
4         4  Community Volunteer      22      False     True     False   
5         4  Community Volunteer      24      False     True      True   
6         4  Community Volunteer      26       True     True     False   
7         4  Community Volunteer     108      False     True     False   
8         4  Community Volunteer     217      False     True      True   
9         4  Community Volunteer     243       True     True      True   
10        4  Community Volunteer      10      False    False     False   
11        4  Community Volunteer      13      False    False     False   
12        4  Community Volunteer      19      False    False     False   
13        4  Community Volunteer      67      False    False     False   
14        4  Community Volunteer      72       True    False     False   
15        4  Community Volunteer      75      False    False     False   
16        4  Community Volunteer     119      False    False     False   
17        4  Community Volunteer     200      False    False     False   
18        4  Community Volunteer     215      False    False     False   
19        4  Community Volunteer     242      False    False     False   

   reactions                                  preprocessed_post  
0          👍  Hi there! Student Affairs is urgently looking ...  
1         ❤️  Hi there! Want any of these? Student Affairs a...  
2         ❤️  We invite you to become an IU Ambassador and t...  
3         ❤️  Volunteering Opportunity at Project management...  
4          🔥  MEGA shopping centre is organizing a charitabl...  
5         ❤️  Dear Students! Soon it's your day! Traditional...  
6          🔥  Volunteering Opportunity: IGD Con Inno GGJ 202...  
7          👍  Dear friends! Art-therapy community invites yo...  
8          🔥  Student Union are looking for amazing VOLUNTEE...  
9          🔥  A few days left before the end of registration...  
10         👎  Want to upgrade your programming skills? We ar...  
11         😐  APPLY for the Robotic projects until December ...  
12         😐  "Program the Future" contest from GS Labs rese...  
13         😐  AMC Makeathon 2020 - Develop applications to s...  
14         👎  Internal ICPC tournament at Innopolis Universi...  
15         😐  REGISTRATION for the BREAKPOINT20 forum is OPE...  
16         👎  Rosneft IT-competition Marathon - Machine Lear...  
17         😐  Red Bull is looking for someone who wants to c...  
18         👎  Online International Conference on Artificial ...  
19         😐  Accounts Chamber of Russian Federation with Ge...

user_id                   description post_id  commented  upvoted  \
0         5  Robotics & Hardware Tinkerer      13      False     True   
1         5  Robotics & Hardware Tinkerer      17      False     True   
2         5  Robotics & Hardware Tinkerer      28       True     True   
3         5  Robotics & Hardware Tinkerer      43       True     True   
4         5  Robotics & Hardware Tinkerer      62       True     True   
5         5  Robotics & Hardware Tinkerer     111      False     True   
6         5  Robotics & Hardware Tinkerer     115      False     True   
7         5  Robotics & Hardware Tinkerer     127      False     True   
8         5  Robotics & Hardware Tinkerer     222      False     True   
9         5  Robotics & Hardware Tinkerer     309      False     True   
10        5  Robotics & Hardware Tinkerer       3       True    False   
11        5  Robotics & Hardware Tinkerer       5       True    False   
12        5  Robotics & Hardware Tinkerer      14      False    False   
13        5  Robotics & Hardware Tinkerer      22      False    False   
14        5  Robotics & Hardware Tinkerer      34       True    False   
15        5  Robotics & Hardware Tinkerer      45      False    False   
16        5  Robotics & Hardware Tinkerer      51      False    False   
17        5  Robotics & Hardware Tinkerer      76      False    False   
18        5  Robotics & Hardware Tinkerer     114      False    False   
19        5  Robotics & Hardware Tinkerer     251       True    False   

    attended reactions                                  preprocessed_post  
0       True         👍  APPLY for the Robotic projects until December ...  
1      False         🔥  20-22 December. Moscow. ML, VRAR, IoT, RPA, Ro...  
2       True         👍  Low Level Programming Club presents a brand-ne...  
3      False         👍  Nintendo Entertainment System, originally know...  
4      False        ❤️  All-Russian MEGAHACKATHON HackUniversity 2020 ...  
5      False         👍  James Dyson Award 2020 James Dyson is on the h...  
6       True        ❤️  Innopolis Open in Robotics (online), 19-21 Jun...  
7       True        ❤️  ARVR Hackathon in arts culture at Kazan State ...  
8       True         👍  1st workshop on electronics and 3D design basi...  
9       True         👍  Robosoft vacancy: Assistant Manager. Full-time...  
10     False         👎  Bonjour! Ça va? С'est la vie.. Croissant. If t...  
11     False         👎  Ho-Ho-Ho! SECRET SANTA IS COMING SOON! Secret ...  
12     False         👎  ART- Therapy course! Dear students! We continu...  
13     False         😐  MEGA shopping centre is organizing a charitabl...  
14     False         😐  Japanese language courses at Anime Club! The f...  
15     False         👎  Poetry evening: date and location change! 16 F...  
16     False         😐  SPRING BALL 2020 When: 2nd April 19:00 What: E...  
17     False         😐  International Fest - is a celebration of cultu...  
18     False         👎  We are the World media contest to address the ...  
19     False         👎  We invite you to the series of lectures on the...

user_id      description post_id  commented  upvoted  attended reactions  \
9         6  Creative Person     834      False     True      True         🥰   
6         6  Creative Person    1705       True     True      True         👍   
4         6  Creative Person    1999      False     True      True        ❤️   
2         6  Creative Person    2064       True     True      True        ❤️   
12        6  Creative Person    2206       True     True     False         🥰   
0         6  Creative Person    2499       True     True      True         😇   
5         6  Creative Person    2634       True     True      True         😇   
11        6  Creative Person    2699      False     True      True         👍   
10        6  Creative Person     364       True     <NA>      True        ❤️   
8         6  Creative Person     872       True     <NA>     False         😇   
7         6  Creative Person    1244      False     <NA>      True         🔥   
3         6  Creative Person    2187      False     <NA>      True         🥰   
1         6  Creative Person    2403      False     <NA>      True         🥰   

                                    preprocessed_post  
9   Second Art Therapy meeting will happen this We...  
6   Welcome to InnoPhotoClub! InnoPhotoClub - a co...  
4   ELECTRINNO - live electronic music Live electr...  
2   Acoustic Campfire Guitar Evenings club invites...  
12  A photo zone is also waiting for you. Come tak...  
0   Guitar Evenings club invites everyone to celeb...  
5   International Art Exhibition This exhibition i...  
11  Photo contest for the best photos of cultural ...  
10  InnoMedia Club are looking for new members of ...  
8   Are you a poet? Or an artist? Want to share yo...  
7   Dear friends! We revive our Art-therapy meetin...  
3   Music concert dedicated to Knowledge Day Join ...  
1   RAGE Club Poetry Evening is coming Hey there, ...

user_id  description post_id  commented  upvoted  attended reactions  \
5         7  Sports User     155      False     True     False         🔥   
4         7  Sports User    1265      False     True      True         🔥   
1         7  Sports User    2045      False     True      True         👍   
12        7  Sports User    2481      False     True      True         🥰   
6         7  Sports User      52       True     <NA>     False         😱   
3         7  Sports User     385      False     <NA>      True        ❤️   
2         7  Sports User     639      False     <NA>      True         👍   
8         7  Sports User     885      False     <NA>      True         🔥   
11        7  Sports User    1258       True     <NA>      True         🥰   
10        7  Sports User    1487      False     <NA>      True         🔥   
7         7  Sports User    1982      False     <NA>      True        ❤️   
13        7  Sports User    2035      False     <NA>      True         🔥   
0         7  Sports User    2418       True     <NA>      True         👍   
9         7  Sports User    2738      False     <NA>      True        ❤️   
15        7  Sports User    2775      False     <NA>      True         🥰   
14        7  Sports User    2823      False     <NA>      True         🔥   

                                    preprocessed_post  
5   IU Sports Festival Table tennis competition! 1...  
4   On March 25th, there will be a city table tenn...  
1   Want to experience the thrill of team play, bo...  
12  Quiz Night is Tonight! We waiting you to our B...  
6   Sports Festival has started last weekend!!! Af...  
3   Volleyball Tournament as part of IU Sports Fes...  
2   Sports Festival: Volleyball Tournament Innopol...  
8   Summer Football Tournament The semester is alm...  
11  Football Quiz Be a captain and gather your tea...  
10  Sports Day in Innopolis: Fitness Fest on Augus...  
7   Free football tickets for students Do you want...  
13  Birthday pre-party for homies On the eve of th...  
0   Innopolis University Wrestling Competition Gre...  
9   KABANIY MARATHON May 4 Innopolis Innopolis 42 ...  
15  Ak Bars Bank launches an application campaign ...  
14  TechArena Kazan Challenge hackathon on optimiz...

user_id     description post_id  commented  upvoted  attended reactions  \
5         8  Mindful Person    1117      False     True     False        ❤️   
7         8  Mindful Person    1768       True     True      True         😇   
4         8  Mindful Person    1862      False     True      True         😱   
8         8  Mindful Person    1950       True     True     False         🤯   
14        8  Mindful Person    2109       True     True      True         🥰   
3         8  Mindful Person    2243       True     True      True        ❤️   
12        8  Mindful Person    2306       True     True     False        ❤️   
13        8  Mindful Person    2393       True     True      True         🥰   
0         8  Mindful Person    2620       True     True      True         🥰   
11        8  Mindful Person    2797      False     True      True         👍   
9         8  Mindful Person    2812       True     True     False         🥰   
19        8  Mindful Person    1633      False    False     False         😭   
22        8  Mindful Person    2381      False    False     False         😭   
17        8  Mindful Person    2477      False    False     False         😭   
20        8  Mindful Person    2602      False    False     False         👎   
16        8  Mindful Person    2625       True    False     False         👎   
15        8  Mindful Person    2650       True    False     False         😭   
6         8  Mindful Person     341       True     <NA>      True         😭   
21        8  Mindful Person    2444       True     <NA>     False         😭   
18        8  Mindful Person    2452       True     <NA>     False         👎   
1         8  Mindful Person    2460      False     <NA>      True         😇   
2         8  Mindful Person    2466       True     <NA>     False         🤯   
10        8  Mindful Person    2804       True     <NA>      True        ❤️   

                                    preprocessed_post  
5   Miss Tatarstan beauty contest We invite the mo...  
7   Football training for girls GirlPower is the f...  
4   Dear girls! In anticipation of International W...  
8   Girls, welcome to the Sisterhood - Innopolis U...  
14  Dear friends! Tomorrow we have got the next li...  
3   Bachata Lady Style Master Class Social Dance C...  
12  Dear students! We invite you to prepare mental...  
13  Dear friends! Tomorrow we have got a little ar...  
0   Dear girls! We invite you to a cozy and inspir...  
11  Hello, student! We invite you to take part in ...  
9   Dear students! TOMORROW We invite you to take ...  
19  Contour invites you to a Technoquiz! Technoqui...  
22  SmartMed Hackathon SmartMed is a unique event ...  
17  Quiz Night is Almost Here! Celebrate our unive...  
20  VR Hackathon 2025: Create the Future of Virtua...  
16  BigQUIZ by Future Simple! We want to ask you, ...  
15  Reminder: International Quiz 2025 Extravaganza...  
6   Dear Girls, We invite you to a meeting with a ...  
21  Hackathon "Norilsk Nickel: Intelligent Horizon...  
18  Harry Potter Quiz In anticipation of the new y...  
1   A lecture for girls from a gynecologist! Dear ...  
2   High Heels Master Class Social Dance Club invi...  
10  Do you want to go to psychotherapy, but are yo...

user_id description post_id  commented  upvoted  attended reactions  \
8         9    Activist    1666      False     True      True         🥰   
15        9    Activist    2158      False     True      True         😱   
10        9    Activist    2270      False     True     False        ❤️   
3         9    Activist    2325       True     True     False         🥰   
2         9    Activist    2395      False     True      True         👍   
14        9    Activist    2717       True     True     False         🥰   
5         9    Activist    2725      False     True     False        ❤️   
4         9    Activist    2029       True    False      True         😱   
17        9    Activist    2109      False    False     False         👎   
11        9    Activist    2160       True    False      True        ❤️   
16        9    Activist    2393       True    False     False         👎   
7         9    Activist    2513       True    False     False        ❤️   
6         9    Activist    2545       True    False      True         🥰   
1         9    Activist    2677       True    False      True         👍   
19        9    Activist    2797      False    False     False         👎   
18        9    Activist    2804       True    False     False         👎   
12        9    Activist    2098       True     <NA>     False         😱   
9         9    Activist    2343       True     <NA>      True        ❤️   
0         9    Activist    2749      False     <NA>     False        ❤️   
13        9    Activist    2829      False     <NA>      True         🥰   

                                    preprocessed_post  
8   Volunteers wanted: INNO GOT TALENT We are look...  
15  CHARITY FITNESS RUN IN INNOPOLIS! Volunteers n...  
10  Volunteer Opportunities in Science! We invite ...  
3   Антон Скударнов, выпускник УИ 2020 года, прове...  
2   Don't miss the upcoming "Formula of the Future...  
14  Dear Friends! We are looking for someone to he...  
5   We invite our volunteers to join the organizat...  
4   Young Con festival for students by Yandex Yand...  
17  Dear friends! Tomorrow we have got the next li...  
11  We need volunteers for the INNO BOOT CAMP 2024...  
16  Dear friends! Tomorrow we have got a little ar...  
7   Scientific volunteering Scientific volunteerin...  
6   We are looking for cupid volunteers for the fo...  
1   XII Online Forum for Education and Socio-Cultu...  
19  Hello, student! We invite you to take part in ...  
18  Do you want to go to psychotherapy, but are yo...  
12  Call for volunteers to the event Game Open Lab...  
9   Calling all volunteers! Do you want to scare e...  
0   reminder TOMORROW: JOB FAIR 2025 Main Hall, 1s...  
13  Call for wonderful brosis for the Fellowship p...

user_id          description post_id  commented  upvoted  attended  \
7        10  Club & Party Person     231       True     True     False   
6        10  Club & Party Person     618      False     True      True   
19       10  Club & Party Person     994      False     True     False   
3        10  Club & Party Person    1022       True     True      True   
17       10  Club & Party Person    1574       True     True     False   
16       10  Club & Party Person    1843       True     True     False   
0        10  Club & Party Person    1915       True     True      True   
9        10  Club & Party Person    2035      False     True     False   
11       10  Club & Party Person    2657      False     True     False   
14       10  Club & Party Person    2658       True     True     False   
18       10  Club & Party Person    1205      False    False     False   
2        10  Club & Party Person    1394      False    False     False   
15       10  Club & Party Person    2043      False    False     False   
10       10  Club & Party Person    2081       True    False      True   
5        10  Club & Party Person    2793       True    False     False   
12       10  Club & Party Person    2819       True    False      True   
4        10  Club & Party Person     877      False     <NA>     False   
1        10  Club & Party Person    1426      False     <NA>      True   
8        10  Club & Party Person    1488      False     <NA>      True   
13       10  Club & Party Person    2799      False     <NA>     False   

   reactions                                  preprocessed_post  
7          👍  First student StandUp Club Show in Bar 108 in ...  
6          🔥  StandUp Show in Innopolis: ENTER FOR FREE! Fir...  
19         👎  CLUB FEST 2022, TOMORROW Join the largest expo...  
3          🔥  New stand-up show in ArtSpace Starring: Lev Ma...  
17         👎  All-Russian festival of university technology ...  
16         😭  The International Fest is starting its movemen...  
0          🔥  Inno Stand Up Club invites fellow residents of...  
9         ❤️  Birthday pre-party for homies On the eve of th...  
11         🔥  PARTY Do you remember how cool we lit up at th...  
14         👎  International Fest: important knowledge about ...  
18         👎  INTERNATIONAL FEST 2023 International Fest - i...  
2          🔥  Standup show from Inno Stand Up Club! Comedian...  
15         👎  The sixth festival of young multinational art ...  
10         🔥  Pulsating beats. Dark lights. Real energy БОЙЛ...  
5         ❤️  Inno Stand Up Club presents: "Session" a stand...  
12         🔥  Slippers Party 2025: The Most Comfortable Even...  
4         ❤️  Get your tickets for the Standup Show on May 1...  
1          👍  Standup show from Inno Stand Up Club! Comedian...  
8         ❤️  We have cool news - Innopolis city is going to...  
13         😭  Alfa-Dvizh Festival in Kazan Alfa-Dvizh is a f...

### 🔢 **Create User Profiles** 🔢

In [11]:
from typing import Dict, List, Optional, TypedDict, Union, Tuple
import numpy as np
from enum import Enum
from sentence_transformers import SentenceTransformer
import torch
import gc
import time

2025-07-09 15:06:32.613705: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1752073592.839754      36 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1752073592.899265      36 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [12]:
model_name = "sentence-transformers/paraphrase-multilingual-mpnet-base-v2"

In [13]:
gc.collect()
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats(0)
model_full = SentenceTransformer(
    model_name,
    device="cuda:0" if torch.cuda.is_available() else "cpu",
)
allocated_mb_full = torch.cuda.max_memory_allocated(0) / (1024**2)
reserved_mb_full = torch.cuda.max_memory_reserved(0) / (1024**2)
print("Full Precision")
print(f"- Peak VRAM Allocated: {allocated_mb_full:.2f} MB")
print(f"- Peak VRAM Reserved: {reserved_mb_full:.2f} MB")
del model_full

gc.collect()
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats(0)
model_half = SentenceTransformer(
    model_name,
    model_kwargs={"torch_dtype": torch.float16},
    device="cuda:0" if torch.cuda.is_available() else "cpu",
)
allocated_mb_half = torch.cuda.max_memory_allocated(0) / (1024**2)
reserved_mb_half = torch.cuda.max_memory_reserved(0) / (1024**2)
print("Half Precision")
print(f"- Peak VRAM Allocated: {allocated_mb_half:.2f} MB")
print(f"- Peak VRAM Reserved: {reserved_mb_half:.2f} MB")
del model_half

gc.collect()
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats(0)

print("\n\n")
print(f"Full (fp32): {100*allocated_mb_full / allocated_mb_full:.2f}% (Allocated {allocated_mb_full:.2f} MB, Reserved: {reserved_mb_full:.2f} MB)")
print(f"Half (fp16): {100*allocated_mb_half / allocated_mb_full:.2f}% (Allocated {allocated_mb_half:.2f} MB, Reserved: {reserved_mb_half:.2f} MB)")

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/723 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/402 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Full Precision
- Peak VRAM Allocated: 1060.73 MB
- Peak VRAM Reserved: 1116.00 MB
Half Precision
- Peak VRAM Allocated: 538.00 MB
- Peak VRAM Reserved: 550.00 MB
Full (fp32): 100.00% (Allocated 1060.73 MB, Reserved: 1116.00 MB)
Half (fp16): 50.72% (Allocated 538.00 MB, Reserved: 550.00 MB)


In [14]:
model_full = SentenceTransformer(model_name)
model_half = SentenceTransformer(
    model_name,
    model_kwargs={"torch_dtype": torch.float16},
)

model_full.encode(df_posts['preprocessed_post'].iloc[0])
model_half.encode(df_posts['preprocessed_post'].iloc[0])

start = time.time()
embeddings_full = model_full.encode(df_posts['preprocessed_post'].to_numpy())
full_time = time.time() - start

start = time.time()
embeddings_half = model_half.encode(df_posts['preprocessed_post'].to_numpy())
half_time = time.time() - start

print(f"Full: {1000*full_time:.2f}ms\nHalf: {1000*half_time:.2f}ms")

df_posts['embedding_full'] = list(np.array(embeddings_full))
df_posts['embedding_half'] = list(np.array(embeddings_half))
df_posts.head()

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/90 [00:00<?, ?it/s]

Batches:   0%|          | 0/90 [00:00<?, ?it/s]

Full: 9053.31ms
Half: 8710.69ms


post_id                                               post  \
0       0  Dear students,\n\nThis channel advertises mino...   
1       1  📣Hi there!\n\nStudent Affairs is urgently look...   
2       2  📣Hi there! Want any of these?\n\nStudent Affai...   
3       3  Bonjour! Ça va?\nС'est la vie.. \nCroissant.\n...   
4       4  📣On December 14, comedian Vladimir Marconi arr...   

                                   preprocessed_post embedding  \
0  Dear students, This channel advertises minor e...      None   
1  Hi there! Student Affairs is urgently looking ...      None   
2  Hi there! Want any of these? Student Affairs a...      None   
3  Bonjour! Ça va? С'est la vie.. Croissant. If t...      None   
4  On December 14, comedian Vladimir Marconi arri...      None   

                                      embedding_full  \
0  [-0.0035118628, -0.083923094, -0.003956843, 0....   
1  [-0.037539687, 0.036673505, -0.0051367693, -0....   
2  [-0.072191216, 0.13000107, -0.0036890064, 0.00...   
3  [-0.008441643, -0.05099034, -0.006417022, 0.04...   
4  [-0.07568409, 0.28184435, -0.004301307, -0.054...   

                                      embedding_half  
0  [-0.00347, -0.08386, -0.00396, 0.03067, 0.1498...  
1  [-0.03748, 0.0367, -0.00514, -0.02245, 0.1855,...  
2  [-0.0722, 0.1301, -0.003689, 0.009346, 0.1477,...  
3  [-0.00841, -0.05096, -0.006416, 0.04224, 0.001...  
4  [-0.0756, 0.282, -0.004307, -0.05408, 0.1282, ...

In [15]:
class ReactionType(Enum):
    GOOD = "good"
    NEUTRAL = "neutral"
    BAD = "bad"

df_emoji = pd.DataFrame(
    columns=['emoji', 'interpretation']
)


def converted_emoji(emoji: str) -> List[ReactionType]:
    match = df_emoji[df_emoji['emoji'] == emoji]['interpretation']
    if not match.empty:
        return [match.iloc[0]]
    return [None]

df_emoji.loc[len(df_emoji)] = ['❤️', ReactionType.GOOD]
df_emoji.loc[len(df_emoji)] = ['🥰', ReactionType.GOOD]
df_emoji.loc[len(df_emoji)] = ['😱', ReactionType.NEUTRAL]
df_emoji.loc[len(df_emoji)] = ['👍', ReactionType.GOOD]
df_emoji.loc[len(df_emoji)] = ['👎', ReactionType.BAD]
df_emoji.loc[len(df_emoji)] = ['😇', ReactionType.NEUTRAL]
df_emoji.loc[len(df_emoji)] = ['😭', ReactionType.NEUTRAL]
df_emoji.loc[len(df_emoji)] = ['🔥', ReactionType.GOOD]
df_emoji.loc[len(df_emoji)] = ['😐', ReactionType.BAD]

df_emoji

emoji        interpretation
0    ❤️     ReactionType.GOOD
1     🥰     ReactionType.GOOD
2     😱  ReactionType.NEUTRAL
3     👍     ReactionType.GOOD
4     👎      ReactionType.BAD
5     😇  ReactionType.NEUTRAL
6     😭  ReactionType.NEUTRAL
7     🔥     ReactionType.GOOD
8     😐      ReactionType.BAD

#### **Steps:**
1. ```EventInteraction```: Typed dictionary defining user interactions (commented, upvoted, attended, reactions).
2. ```UserProfileBuilder```: 
   - Computes interaction weights using predefined weights
   - Builds three embeddings:
        good_profile: Weighted average of positive-interaction event embeddings
        bad_profile: Weighted average of negative-interaction event embeddings
        combined_profile: good_profile - λ*bad_profile
3. ```Processing```:
   - Groups user interactions by user_id
   - For each user:
        a. Collects all interactions and corresponding post embeddings
        b. Computes combined user profile embedding
   - Returns DataFrame with user_id and profile embedding

In [16]:
class EventInteraction(TypedDict):
    commented: bool
    upvoted: Optional[bool]
    attended: bool
    reactions: List[ReactionType]

class UserProfileBuilder:
    def __init__(self, weights_config: Optional[Dict] = None, lambda_param = 1):
        self.default_weights = {
            "commented": 1.3,
            "upvoted": {
                True: 2,     # Upvote
                False: -0.5, # Downvote
                None: 1      # No reaction
            },
            "attended": 2.0,
            "reactions": {
                ReactionType.GOOD: 1.2,
                ReactionType.NEUTRAL: 1,
                ReactionType.BAD: -0.8
            }
        }
        self.lambda_param = lambda_param
        self.weights = weights_config or self.default_weights

    def compute_weight(self, interaction: EventInteraction) -> float:
        total_weight = 0.0
        
        if interaction["commented"]:
            total_weight += self.weights["commented"]
        
        vote_key = interaction["upvoted"]
        total_weight += self.weights["upvoted"].get(vote_key, 0.0)

        if interaction["attended"]:
            total_weight += self.weights["attended"]
        
        for reaction in interaction["reactions"]:
            total_weight += self.weights["reactions"].get(reaction, 0.0)
            
        return total_weight

    def build_profile(
        self,
        interactions: List[EventInteraction], 
        event_embeddings: List[np.ndarray],
        return_components: bool = False
    ) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray, np.ndarray]]:
        assert len(interactions) == len(event_embeddings)
        
        good_sum = np.zeros_like(event_embeddings[0])
        good_total_weight = 0.0
        bad_sum = np.zeros_like(event_embeddings[0])
        bad_total_weight = 0.0

        for i, emb in enumerate(event_embeddings):
            weight = self.compute_weight(interactions[i])
            
            if weight > 0:
                good_sum += weight * emb
                good_total_weight += weight
            elif weight < 0:
                abs_weight = abs(weight)
                bad_sum += abs_weight * emb
                bad_total_weight += abs_weight
        
        good_profile = good_sum / good_total_weight if good_total_weight > 0 else good_sum
        bad_profile = bad_sum / bad_total_weight if bad_total_weight > 0 else bad_sum
        
        combined_profile = good_profile - self.lambda_param * bad_profile
        
        combined_profile_norm = np.linalg.norm(combined_profile)
        if combined_profile_norm > 0:
            combined_profile /= combined_profile_norm
        
        if return_components:
            return combined_profile, good_profile, bad_profile
        
        return combined_profile


In [17]:
grouped = df_users.groupby('user_id')
for user_id, group in grouped:
    for _, row in group.iterrows():
        post_emb = df_posts[df_posts['post_id'] == row['post_id']]['embedding']
        print(row)
        print(post_emb)
        if post_emb is None:
            continue
        post_emb = post_emb.iloc[0]
        # print(post_emb)
        break
    break

user_id                        1
description    Competitive Coder
post_id                       10
commented                  False
upvoted                     True
attended                    True
reactions                     ❤️
Name: 0, dtype: object
10    None
Name: embedding, dtype: object


In [18]:
user_profiles = {}

builder = UserProfileBuilder(lambda_param=0.6)
results = []
grouped = df_users.groupby('user_id')

for user_id, group in grouped:
    interactions_list = []
    embeddings_list_full = []
    embeddings_list_half = []
    
    for _, row in group.iterrows():
        interaction = EventInteraction(
            commented=row['commented'],
            upvoted=row['upvoted'],
            attended=row['attended'],
            reactions=converted_emoji(row['reactions'])
        )
        
        post_emb_full = df_posts[df_posts['post_id'] == row['post_id']]['embedding_full']
        post_emb_half = df_posts[df_posts['post_id'] == row['post_id']]['embedding_half']
        
        if (post_emb_full is None) or (post_emb_half is None):
            continue
        post_emb_full = post_emb_full.iloc[0]
        post_emb_half = post_emb_half.iloc[0]
        
        interactions_list.append(interaction)
        embeddings_list_full.append(post_emb_full)
        embeddings_list_half.append(post_emb_half)

    combined_emb_full = None
    combined_emb_half = None
    if interactions_list:
        combined_emb_full = builder.build_profile(
            interactions_list, 
            embeddings_list_full
        )
        combined_emb_half = builder.build_profile(
            interactions_list, 
            embeddings_list_half
        )
        user_profiles[user_id] = combined_emb_full
    
    results.append({
        'user_id': user_id,
        'profile_full': combined_emb_full,
        'profile_half': combined_emb_half
    })

df_user_profiles = pd.DataFrame(results)
df_user_profiles

user_id                                       profile_full  \
0        1  [-0.04450846, 0.033089314, 0.0007838856, 0.007...   
1        2  [0.00124734, 0.0863814, -0.0017737267, 0.00921...   
2        3  [-0.041098606, -0.00010002946, -0.0007372631, ...   
3        4  [0.0008095659, -0.00982608, -0.0037328796, 0.0...   
4        5  [0.013821459, -0.008949878, 0.00036891518, -0....   
5        6  [-0.03872987, 0.010225302, -0.002607824, 0.016...   
6        7  [-0.030038191, 0.025484309, -0.0024657212, 0.0...   
7        8  [-0.025541296, 0.017996315, -0.0012445541, 0.0...   
8        9  [-0.025503308, 0.018868852, -0.0026850773, -0....   
9       10  [-0.025794191, 0.04356196, -0.00094257045, -0....   

                                        profile_half  
0  [-0.04446, 0.0331, 0.000784, 0.006996, 0.00553...  
1  [0.001251, 0.0863, -0.001774, 0.00924, -0.0266...  
2  [-0.0411, -0.0001445, -0.0007405, -0.0294, 0.0...  
3  [0.000821, -0.00983, -0.00373, 0.00787, 0.0373...  
4  [0.01385, -0.008995, 0.0003724, -0.01046, 0.05...  
5  [-0.03867, 0.01021, -0.00261, 0.0169, 0.001943...  
6  [-0.03, 0.02547, -0.002468, 0.03668, 0.0357, -...  
7  [-0.02554, 0.01799, -0.001242, 0.02802, 0.0114...  
8  [-0.02545, 0.0189, -0.002687, -0.01797, 0.0346...  
9  [-0.02574, 0.0437, -0.0009418, -0.0633, 0.0138...

# ⭐ **Create recommendations** ⭐

In [19]:
from sklearn.metrics.pairwise import cosine_similarity
from typing import List

In [20]:
def get_recommendations(user_id: int, k: int = 5, mode: str = "full") -> List[int]:
    if mode not in ("full", "half"):
        raise ValueError()

    user_profile_vector = df_user_profiles.loc[df_user_profiles["user_id"] == user_id, f"profile_{mode}"].iloc[0]
    
    seen_post_ids = df_users.loc[df_users["user_id"] == user_id, "post_id"].unique()
    unseen_posts = df_posts[~df_posts["post_id"].isin(seen_post_ids)].copy()

    all_post_embeddings = np.vstack(unseen_posts[f"embedding_{mode}"].values)

    scores = cosine_similarity(
        user_profile_vector.reshape(1, -1),
        all_post_embeddings
    ).flatten()
    unseen_posts["score"] = scores

    recommended_posts = unseen_posts.sort_values(by="score", ascending=False)

    return recommended_posts.head(k)["post_id"].tolist()

# 📈 **Evaluate recommended documents relevancy** 📈

In [27]:
import numpy
from typing import Set
from numpy.typing import NDArray

In [28]:
user_keywords_dict = {
    1: {"icpc", "hackathon", "ctf", "makeathon", "datacontest", "timus", "programming", "developer", "contest", "challenge", "gamedev", "algorithm", "cybersecurity", "technical"},
    2: {"art therapy", "poetry evening", "spring ball", "dance festival", "anime club", "concert", "sketch", "art", "creative", "music", "dance", "design", "exhibition", "performance", "culture", "drawings"},
    3: {"internship", "vacancy", "career", "soft skills", "job fair", "trainee", "recruiting", "business", "professional", "development", "project", "forum", "skills", "grant", "startup"},
    4: {"volunteer", "volunteering", "ambassador", "charity", "student union", "handouts", "community", "social", "help", "support", "affairs", "assistants", "duties"},
    5: {"robotics", "hardware", "electronics", "iot", "dyson award", "arvr", "3d design", "actuators", "engineering", "develop", "workshop", "robotic", "robot", "device", "technical", "gamedev"},
    6: {"concert", "stand up", "party", "disco", "collab", "artspace", "guitar evenings", "music", "creative", "show", "performance", "dj", "band", "art", "quiz", "artists"},
    7: {"football", "cricket", "volleyball", "tennis", "basketball", "triathlon", "sports festival", "frisbee", "sports", "tournament", "competition", "game", "team", "club", "match", "fitness"},
    8: {"art", "therapy", "psychology", "psychological", "gynecologist", "sexologist", "sisterhood", "health", "well-being", "stress", "emotions", "anxiety", "self-care", "support", "girls"},
    9: {"volunteer", "volunteering", "ambassador", "fellowship", "mentor", "charity", "community", "project", "forum", "hackathon", "olympiad", "support", "career", "development"},
    10: {"stand up", "party", "artspace", "pre-party", "disco", "bar 108", "open mic", "club", "show", "comedy", "comedians", "evening", "dj", "night", "concert", "performers"}
}

In [29]:
def get_relevant_items(user_id: int, keywords: set, threshold: int = 5) -> Set[int]:
    seen_post_ids = df_users.loc[df_users["user_id"] == user_id, "post_id"].unique()
    unseen_posts = df_posts[~df_posts["post_id"].isin(seen_post_ids)].copy()

    if unseen_posts.empty:
        return set()

    match_counts = np.zeros(len(unseen_posts), dtype=int)
    for keyword in keywords:
        keyword_present = unseen_posts["preprocessed_post"].str.contains(keyword, case=False, na=False)
        match_counts += keyword_present.to_numpy()

    relevancy_mask = match_counts >= threshold
    relevant_ids = set(unseen_posts.loc[relevancy_mask, "post_id"])

    return relevant_ids


def precision_at_k(relevance: NDArray[np.int64], k: int) -> float:
    relevance = relevance[:k]
    return np.sum(relevance) / k if k > 0 else 0.0


def average_precision_at_k(relevance: NDArray[np.int64], k: int) -> float:
    relevance = relevance[:k]
    indices = np.arange(1, k + 1)
    cumulative_hits = np.cumsum(relevance)
    precision_at_hits = cumulative_hits[relevance == 1] / indices[relevance == 1]
    return np.mean(precision_at_hits) if np.sum(relevance) > 0 else 0.0


def reciprocal_rank(relevance: NDArray[np.int64]) -> float:
    first_hit_index = np.argmax(relevance)
    if relevance[first_hit_index] == 1:
        return 1 / (first_hit_index + 1)
    return 0.0


def ndcg_at_k(relevance: NDArray[np.int64], k: int) -> float:
    relevance = relevance[:k]
    positions = np.arange(1, k + 1)
    discounts = np.log2(positions + 1)
    dcg = np.sum(relevance / discounts)
    ideal_relevance = np.sort(relevance)[::-1]
    idcg = np.sum(ideal_relevance / discounts)
    return dcg / idcg if idcg > 0 else 0.0

In [30]:
K_VALUE = 50
all_metrics = []

for user_id in range(1, 11):
    keywords = user_keywords_dict[user_id]
    relevant_ids = get_relevant_items(user_id=user_id, keywords=keywords, threshold=2)
    recommendations = get_recommendations(user_id=user_id, k=K_VALUE, mode="full")
    
    relevance = np.array([1 if post_id in relevant_ids else 0 for post_id in recommendations])
    user_metrics = {
        "user_id": user_id,
        f"Precision@{K_VALUE}": precision_at_k(relevance, K_VALUE),
        f"AP@{K_VALUE}": average_precision_at_k(relevance, K_VALUE),
        "RR": reciprocal_rank(relevance),
        f"NDCG@{K_VALUE}": ndcg_at_k(relevance, K_VALUE)
    }
    all_metrics.append(user_metrics)

df_final_evaluation_full = pd.DataFrame(all_metrics)
print(f"=== Final Evaluation Metrics @{K_VALUE} (Full Precision) ===")
display(df_final_evaluation_full)

=== Final Evaluation Metrics @50 (Full Precision) ===


user_id  Precision@50     AP@50   RR   NDCG@50
0        1          0.50  0.694691  1.0  0.909424
1        2          0.40  0.496314  1.0  0.807361
2        3          0.40  0.446589  0.5  0.733221
3        4          0.44  0.636813  1.0  0.882046
4        5          0.62  0.739056  1.0  0.922462
5        6          0.74  0.716966  1.0  0.909882
6        7          0.74  0.772263  1.0  0.932620
7        8          0.54  0.802848  1.0  0.949359
8        9          0.54  0.593807  1.0  0.856456
9       10          1.00  1.000000  1.0  1.000000

In [31]:
K_VALUE = 50
all_metrics = []

for user_id in range(1, 11):
    keywords = user_keywords_dict[user_id]
    relevant_ids = get_relevant_items(user_id=user_id, keywords=keywords, threshold=2)
    recommendations = get_recommendations(user_id=user_id, k=K_VALUE, mode="half")
    
    relevance = np.array([1 if post_id in relevant_ids else 0 for post_id in recommendations])
    user_metrics = {
        "user_id": user_id,
        f"Precision@{K_VALUE}": precision_at_k(relevance, K_VALUE),
        f"AP@{K_VALUE}": average_precision_at_k(relevance, K_VALUE),
        "RR": reciprocal_rank(relevance),
        f"NDCG@{K_VALUE}": ndcg_at_k(relevance, K_VALUE)
    }
    all_metrics.append(user_metrics)

df_final_evaluation_half = pd.DataFrame(all_metrics)
print(f"=== Final Evaluation Metrics @{K_VALUE} (Hald Precision) ===")
display(df_final_evaluation_half)

=== Final Evaluation Metrics @50 (Hald Precision) ===


user_id  Precision@50     AP@50   RR   NDCG@50
0        1          0.50  0.694691  1.0  0.909424
1        2          0.40  0.496314  1.0  0.807361
2        3          0.40  0.448104  0.5  0.734458
3        4          0.44  0.636813  1.0  0.882046
4        5          0.62  0.739056  1.0  0.922462
5        6          0.74  0.716966  1.0  0.909882
6        7          0.74  0.772263  1.0  0.932620
7        8          0.54  0.804170  1.0  0.949760
8        9          0.54  0.581461  1.0  0.841143
9       10          1.00  1.000000  1.0  1.000000

In [34]:
metrics = df_final_evaluation_full.columns
results_avg = {}
for metric in metrics:
    if metric == "user_id":
        continue
    metric_full = numpy.mean(df_final_evaluation_full[metric])
    metric_half = numpy.mean(df_final_evaluation_half[metric])
    results_avg[metric] = 100 * ((metric_half - metric_full)/ metric_full)

for res, delta in results_avg.items():
    print(f"{res}: {delta:.2f}%")

Precision@50: 0.00%
AP@50: -0.14%
RR: 0.00%
NDCG@50: -0.15%
